### ISSUE TITLE
Multichip Hang With Watcher Assert

### ISSUE THREAD

**Issue by SeanNijjar** [tenstorrent-metal/tt-metal#6363]
Run the following command on a T3000:
`pytest tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py -k "post_commit and looping and  200-mem_config1-input_dtype1-8-1-input_shape2-3-layout2"`

from branch `snijjar/issue-6363`

You'll hit this error:

```
2024-03-13 16:37:39.918 | INFO     | tests.tt_eager.python_api_testing.unit_testing.misc.test_all_gather:run_all_gather_on_t3000_impl:99 - dim: 3
              LLRuntime | INFO     | Watcher checking device 3
              LLRuntime | INFO     | Watcher checking device 2
              LLRuntime | INFO     | Watcher checking device 6
                 Always | WARNING  | Watcher stopped the device due to bad NOC unicast transaction
                 Always | INFO     | While running kernels:
                 Always | INFO     |  brisc : tt_metal/impl/dispatch/kernels/cq_prefetcher.cpp
                 Always | INFO     |  ncrisc: blank
                 Always | INFO     |  triscs: blank
                 Always | INFO     | Last waypoint: NAWW,W,W,W,W 
                 Always | FATAL    | On core (x=2,y=9): brisc using noc0 tried to accesss core (00,03) L1[addr=0x800000000,len=4]
```

**Comment by jliangTT** [comment]
what is the reason for P0 for this issue? (P0 is pretty disruptive).

**Comment by SeanNijjar** [comment]
@davorchap Suggested I bump up this item since it is blocking all-gather changes from getting into main (and having knock-on affects as a result - model perf, blocked, other feature development, burning dev time, etc.).

**Comment by SeanNijjar** [comment]
I agree P0 is touchy though, so let's chat if there is push back on that.

**Comment by jliangTT** [comment]
i think that is a good reason for me. @davorchap  is right. I agree with it.

**Comment by DrJessop** [comment]
Pushed a fix to address the watcher error. Confirmed that if we synchronize after every op invocation, it passed, but it hangs if we don't block. Can you confirm this workaround works for you?
If so, you can make a new issue to address the new problem and assign to me.

**Comment by DrJessop** [comment]
Opened issue for part 2 of debugging: #6404

I also checked in my fix commit, so maybe see if there is a way for you to merge.

**Comment by aliuTT** [comment]
The hangs we're consistently seeing with program cache enabled on remote chips, is due to bidirectional tunneller not observing the max cmd slots for tensix cores, therefore it can end up in a deadlock scenario where the issue path is trying to forward more cmds when the remote pull and push kernel is blocked trying to send cmds back through the completion path.

With fix `bdc566044`, I am able to run the test here many times, no hangs. I also tried Falcon40B 60 layers with program cache on 8 chips, it's now passing consistently.

The bug can affect any remote chip test (e.g. device 4,5,6,7 on T3000), please try commit with any reported hangs.

**Comment by aliuTT** [comment]
This is fully resolved I believe
